### ISSUE TITLE
What is the best balance between batchSize and maxConcurrentBatchCount ?

### ISSUE THREAD

**Issue by ojm-it** [kafkas/firecode#23]
Let's say that we have a number `K` of docs to traverse. We should have at most `N` docs held in memory at the same time. To get the best performance, should we :
- Choose the maximum `batchSize` possible (=500), and set the `maxConcurrentBatchCount` to `N / batchSize`
- Choose the maximum `maxConcurrentBatchCount` logically possible (which I don't know), and set the `batchSize` to `N / maxConcurrentBatchCount`
- Have preferred values for either the batchSize or the maxConcurrentBatchCount depending on N. In this case, what would these values be ?

Thanks for helping!

**Comment by kafkas** [comment]
Great question! Sorry for the late reply; the solution is a bit involved. First of all, the maximum `batchSize` is not 500. Batches can be as large as you want them to be.

I believe the correct approach is to start from the memory constraint. To be consistent and avoid confusion, M and N will denote the following:

- M: maximum number of docs we want to hold in memory
- N: total number of docs in the traversable

Our primary constraint is

<img src="https://latex.codecogs.com/svg.image?M&space;=&space;batchSize&space;\times&space;maxConcurrentBatchCount" title="M = batchSize \times maxConcurrentBatchCount" />

where M is some constant, or, more precisely, an exogenous variable in this model. Since we have already decided on an upper bound for the space complexity we just need to minimize the time complexity (TC). We know that the time complexity of FastTraverser is

<img src="https://latex.codecogs.com/svg.image?TC&space;=&space;\frac{N}{batchSize}(Q(batchSize)&space;&plus;&space;\frac{C(batchSize)}{maxConcurrentBatchCount})" title="TC = \frac{N}{batchSize}(Q(batchSize) + \frac{C(batchSize)}{maxConcurrentBatchCount})" />

We can substitute <img src="https://latex.codecogs.com/svg.image?\inline&space;\frac{M}{batchSize}" title="\inline \frac{M}{batchSize}" /> for <img src="https://latex.codecogs.com/svg.image?\inline&space;maxConcurrentBatchCount" title="\inline maxConcurrentBatchCount" /> to get

<img src="https://latex.codecogs.com/svg.image?TC&space;=&space;\frac{NQ(batchSize)}{batchSize}&space;&plus;&space;\frac{NC(batchSize)}{M}" title="TC = \frac{NQ(batchSize)}{batchSize} + \frac{NC(batchSize)}{M}" />

Take the derivative of TC with respect to `batchSize`:

<img src="https://latex.codecogs.com/svg.image?\frac{dTC}{dbatchSize}&space;=&space;\frac{-NQ(batchSize)}{batchSize^{2}}&space;&plus;&space;\frac{N{{Q}^\prime}(batchSize)}{batchSize}&space;&plus;&space;\frac{N{{C}^\prime}(batchSize)}{M}" title="\frac{dTC}{dbatchSize} = \frac{-NQ(batchSize)}{batchSize^{2}} + \frac{N{{Q}^\prime}(batchSize)}{batchSize} + \frac{N{{C}^\prime}(batchSize)}{M}" />

Now, to deduce useful information from the analysis we need to have reasonably accurate proxies for the Q and C functions. For simplicity, let's assume that

<img src="https://latex.codecogs.com/svg.image?Q(batchSize)&space;=&space;500&space;&plus;&space;batchSize" title="Q(batchSize) = 500 + batchSize" />

Obviously, this may be completely wrong. The real answer will depend on a variety of factors e.g. your network speed, the size of the documents in the collection etc. but we just need to come up with a good estimate. I believe it’s fair to assume that Q is an increasing function of `batchSize` and using simple linear regression should suffice.

As for the function C, this completely depends on the time complexity of your callback function. It's your responsibility to estimate its processing time. If you're making 100 API calls in the callback, C should reflect that. I think C should also be an increasing function of `batchSize` and actually we only need to know its derivative with respect to `batchSize` (we don't use the function itself). For simplicity, let's assume that

<img src="https://latex.codecogs.com/svg.image?C(batchSize)&space;=&space;1000&space;&plus;&space;batchSize" title="C(batchSize) = 1000 + batchSize" />

To find the local minima we need to find the roots of the equation <img src="https://latex.codecogs.com/svg.image?\inline&space;\frac{dTC}{dbatchSize}&space;=&space;0" title="\inline \frac{dTC}{dbatchSize} = 0" />. So we have

<img src="https://latex.codecogs.com/svg.image?&space;\frac{-N(500&space;&plus;&space;batchSize)}{batchSize^{2}}&space;&plus;&space;\frac{N&space;\times&space;1}{batchSize}&space;&plus;&space;\frac{N&space;\times&space;1}{M}&space;=&space;0" title="&space;\frac{-N(500 + batchSize)}{batchSize^{2}}&space;&plus;&space;\frac{N \times 1}{batchSize}&space;&plus;&space;\frac{N \times 1}{M} = 0" />

You can confirm either by solving this equation manually yourself or by using a tool like Wolfram Alpha (see [this graph](https://www.wolframalpha.com/input/?i=T%28b%29+%3D+%28500+%2B+b%29%2Fb+%2B+%281000+%2B+b%29%2FM+where+M+%3D+10000)) that the time complexity function is a hyperbola with a local minimum at the following point:

<img src="https://latex.codecogs.com/svg.image?batchSize_{0}&space;=&space;\sqrt{500M}\\maxConcurrentBatchCount_{0}&space;=&space;\sqrt{\frac{M}{500}}" title="batchSize_{0} = \sqrt{500M}\\maxConcurrentBatchCount_{0} = \sqrt{\frac{M}{500}}" />

Intuitively, the answer is always a function of your initial memory constraint M. With the above assumptions, if M = 10,000 the optimal `batchSize` is roughly 2236 and `maxConcurrentBatchCount` is roughly 4.

**TL;DR**

To find the optimal values for `batchSize` and `maxConcurrentBatchCount` you need to do the following:

1. Decide what M will be.
2. Decide what <img src="https://latex.codecogs.com/svg.image?\inline&space;Q(batchSize)" title="\inline Q(batchSize)" /> will be.
3. Decide what <img src="https://latex.codecogs.com/svg.image?\inline&space;{C}^\prime(batchSize)" title="\inline {C}^\prime(batchSize)" /> will be.
4. Find the (positive) root of the equation <img src="https://latex.codecogs.com/svg.image?\inline&space;\frac{dTC}{dbatchSize}&space;=&space;0" title="\inline \frac{dTC}{dbatchSize} = 0" /> to get the optimal `batchSize`.
5. Use <img src="https://latex.codecogs.com/svg.image?\inline&space;maxConcurrentBatchCount&space;=&space;\frac{M}{batchSize}" title="\inline maxConcurrentBatchCount = \frac{M}{batchSize}" /> to get the optimal `maxConcurrentBatchCount`.

Hope this helps! Let me know if you have any other questions.

**Comment by ojm-it** [comment]
Thank you for this very detailed answer, it helps a lot !
I've made this small function based on your answer :

```typescript
static getOptimalTraverserConfig(maxNumberOfDocsInMemory: number, maxTraversableSize = Number.POSITIVE_INFINITY): {
        batchSize: number,
        maxConcurrentBatchCount: number,
    } {
        //The max number of docs in memory is capped by the traversable size
        const effective_maxNbOfDocsInMemory = Math.min(maxNumberOfDocsInMemory, maxTraversableSize);

        let maxConcurrentBatchCount = Math.round(Math.sqrt(effective_maxNbOfDocsInMemory / 500));

        if (maxConcurrentBatchCount < 1) {
            maxConcurrentBatchCount = 1;
        }

        const batchSize = Math.floor(effective_maxNbOfDocsInMemory / maxConcurrentBatchCount);

        return {
            batchSize: batchSize,
            maxConcurrentBatchCount: maxConcurrentBatchCount,
        };
    }
```
Given that `maxConcurrentBatchCount` is the the smallest value, to increase precision, we round it first, and then calculate `batchSize` accordingly.

These are the first values of `maxConcurrentBatchCount` :
| maxNumberOfDocsInMemory | maxConcurrentBatchCount |
|:-----------------------:|:-----------------------:|
| < 1125                  |  1                      |
| < 3125                  |  2                      |
| < 6125                  |  3                      |
| < 10125                 |  4                      |
| < 15125                 |  5                      |
| < 21125                 |  6                      |

Do you think that a similar function should be added to Firecode ? I think it's accurate enough for a lot of use cases.

**Comment by kafkas** [comment]
Yes, we can probably provide the users with a similar function that computes the optimal config but I think it would need to take more arguments. We cannot use `Math.sqrt(real_maxNbOfDocsInMemory / 500)` as that is only optimal under the above assumptions. We can probably estimate Q ourselves but C' would probably have to come from the user. Will create a separate issue for this.
